railgun/internal/ctrlutils/ingress-status.go

```diff
@@ -0,0 +1,389 @@
+package ctrlutils
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/kong/deck/file"
+	"github.com/prometheus/common/log"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	knative "knative.dev/networking/pkg/apis/networking/v1alpha1"
+	knativeversioned "knative.dev/networking/pkg/client/clientset/versioned"
+	knativeApis "knative.dev/pkg/apis"
+	"knative.dev/pkg/network"
+
+	"github.com/kong/kubernetes-ingress-controller/pkg/sendconfig"
+	kicclientset "github.com/kong/kubernetes-ingress-controller/railgun/pkg/clientset"
+)
+
+// dedicated function that process ingress/customer resource status update after configuration is updated within kong.
+func PullConfigUpdate(ctx context.Context, kongConfig sendconfig.Kong, log logr.Logger, kubeConfig *rest.Config) {
+	log.Info("Launching Ingress Status Update Thread.")
+	var wg sync.WaitGroup
+	for {
+		select {
+		case updateDone := <-kongConfig.ConfigDone:
+			log.Info("receive configuration information. Update ingress status %v \n", updateDone)
+			wg.Add(1)
+			go UpdateIngress(ctx, &updateDone, log, kubeConfig, &wg)
+			wg.Wait()
+		case <-ctx.Done():
+			log.Info("stop status update channel.")
+			return
```

Comment:
When waiting for a context to complete it is usually a good idea to differentiate whether or not the context had any error associated with it's completion. I suggest we check and log based on error:

```suggestion
			if err := ctx.Err(); err != nil {
				log.Error(err, "status update thread stopped")
			} else {
				log.Info("status update thread stopped: context completed")
			}
			return
```